sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -436,60 +437,94 @@ static size_t _handle_req(coap_pkt_t *pdu, uint8_t *buf, size_t len,
+            continue;
+        }
+        else {
+            return GCOAP_RESOURCE_FOUND;
+        }
+    }
+
+    return ret;
+}
+
 /*
  * Searches listener registrations for the resource matching the path in a PDU.
  *
+ * param[in]  pdu -- the PDU to check the resource for
  * param[out] resource_ptr -- found resource
  * param[out] listener_ptr -- listener for found resource
  * return `GCOAP_RESOURCE_FOUND` if the resource was found,
  *        `GCOAP_RESOURCE_WRONG_METHOD` if a resource was found but the method
  *        code didn't match and `GCOAP_RESOURCE_NO_PATH` if no matching
  *        resource was found.
  */
-static int _find_resource(coap_pkt_t *pdu, const coap_resource_t **resource_ptr,
-                                            gcoap_listener_t **listener_ptr)
+static int _find_resource(const coap_pkt_t *pdu,
+                          const coap_resource_t **resource_ptr,
+                          gcoap_listener_t **listener_ptr)
 {
-    int ret = GCOAP_RESOURCE_NO_PATH;
-    coap_method_flags_t method_flag = coap_method2flag(coap_get_code_detail(pdu));
-
     /* Find path for CoAP msg among listener resources and execute callback. */
     gcoap_listener_t *listener = _coap_state.listeners;
 
-    uint8_t uri[CONFIG_NANOCOAP_URI_MAX];
-    if (coap_get_uri_path(pdu, uri) <= 0) {
-        return GCOAP_RESOURCE_NO_PATH;
-    }
-
     while (listener) {
-        const coap_resource_t *resource = listener->resources;
-        for (size_t i = 0; i < listener->resources_len; i++) {
-            if (i) {
-                resource++;
-            }
+        const coap_resource_t *resource;
+        int res = listener->request_matcher(listener, &resource, pdu);
 
-            int res = coap_match_path(resource, uri);
-            if (res > 0) {
-                continue;
-            }
-            else if (res < 0) {
-                /* resources expected in alphabetical order */
-                break;
-            }
-            else {
-                if (! (resource->methods & method_flag)) {
-                    ret = GCOAP_RESOURCE_WRONG_METHOD;
-                    continue;
-                }
-
-                *resource_ptr = resource;
-                *listener_ptr = listener;
-                return GCOAP_RESOURCE_FOUND;
-            }
+        /* check next resource on mismatch */
+        if (res == GCOAP_RESOURCE_NO_PATH) {
+            listener = listener->next;
+            continue;
         }
-        listener = listener->next;
+        /* found a suitable resource */
+        else if (res == GCOAP_RESOURCE_FOUND) {
+            *resource_ptr = resource;
+            *listener_ptr = listener;
+            return GCOAP_RESOURCE_FOUND;
+        }
+
+        /* res is probably GCOAP_RESOURCE_ERROR or
+         * GCOAP_RESOURCE_NO_PATH. There is an error with
+         * iterating the listener, break. */
+        break;
     }
 
-    return ret;
+    return GCOAP_RESOURCE_NO_PATH;
```

Comment:
This used to return WRONG_METHOD for a PUT to something like /board; can't test while on the road, but looks like it doesn't any more. (It should to preserve that feature and for sake of debugability)

I think the easiest way to do that is to`return res` rather than break. (That would necessitate another case for ERROR -> INTERNAL_SERVER_ERROR like there is ones in 345..348, but that need follows from introducing ERROR in the first place. (Then, on lack of alphabetical sorting, return with error would work too, which I'd accept as out of scope here if you think so, but would work well in here too as the error result is introduced here, and it eases debugging by not failing silently).

Reply:
https://github.com/RIOT-OS/RIOT/pull/14029/commits/359ff57d5ad3da8aaf5df81382bcecf1db7d02ca

this commit adds more error handling.  Wrong methods now correctly produce 4.05 responses. Any other error (I had to enforce that in code for testing) produces a 5.00 internal server error.  I think we need much more refactoring and restructuring of the error cases, but let's handle this in another PR.